src/core/server/http/https_redirect_server.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Request, ResponseToolkit, Server } from 'hapi-latest';
+import { format as formatUrl } from 'url';
+
+import { Logger } from '../logging';
+import { HttpConfig } from './http_config';
+import { createServer, getServerOptions } from './http_tools';
+
+export class HttpsRedirectServer {
+  private server?: Server;
+
+  constructor(private readonly log: Logger) {}
+
+  public async start(config: HttpConfig) {
+    if (!config.ssl.enabled || config.ssl.redirectHttpFromPort === undefined) {
+      throw new Error(
+        'Redirect server cannot be started when [ssl.enabled] is set to `false`' +
+          ' or [ssl.redirectHttpFromPort] is not specified.'
```

Comment:
**question:** does this happen? when do we get into this `start` function but ssl is disabled or redirectHttpFromPort is undefined?

Reply:
You're right, currently there is no code that would call `start` in this case. I was just thinking from position of `HttpsRedirectServer` that doesn't know who's calling it and receives config with __optional__ `ssl.redirectHttpFromPort`.

Generic solution for such case would be passing _defined_ `redirectHttpFromPort` as an argument so that this code can be sure that this is defined, but the problem that this method uses `getServerOptions` under the hood that requires full HttpConfig... Maybe we can redesign this at some point, but I don't have better ideas right now. 

Please tell me what you think would be better to do here and I'll do that.

Reply:
> This https redirect server is for routing requests from the http server to a new https server.

Yes.

> so why would we ignore the TLS part of options in this server?

Cause this server should _only_ accept HTTP traffic, or it's not what you're asking? Also you can take a look at these issues:

https://github.com/elastic/kibana/issues/10142
https://github.com/elastic/kibana/issues/10948
https://github.com/elastic/kibana/pull/10930

Reply:
Thanks for links to issues, read through them to refresh my memory on this.